volttron/platform/auth.py

```diff
@@ -131,6 +147,196 @@ def setup_zap(self, sender, **kwargs):
+        except Unreachable:
+            _log.warn(f'{peer_id} is unreachable while attempting to get rpc methods')
+        except MethodNotFound as e:
+            _log.error(e)
+
+        for method in peer_methods:
+            try:
+                rpc_method_authorizations[method] = \
+                    self.vip.rpc.call(peer_id, 'get_rpc_authorizations', method).get(timeout=4)
+            except gevent.Timeout:
+                rpc_method_authorizations[method] = None
+                _log.warn(f'{peer_id} has timed out while attempting to get rpc method authorizations')
+            except Unreachable:
+                _log.warn(f'{peer_id} is unreachable while attempting to get rpc method authorizations')
+                rpc_method_authorizations[method] = None
+            except MethodNotFound as e:
+                _log.error(e)
+                rpc_method_authorizations[method] = []
+
+        return rpc_method_authorizations
+
+    @RPC.export
+    def update_auth_entry_rpc_method_authorizations(self, identity, rpc_methods):
+        """
+        Update RPC methods for an auth entry. This is called by the subsystem on
+        agent start-up to ensure that the agent's current rpc allowances are recorded
+        with it's auth entry.
+        :param identity: The agent's identity in the auth entry
+        :param rpc_methods: The rpc methods to update, in the format {rpc_method_name: [allowed_rpc_capability_1, ...]}
+        :return: updated_rpc_methods or None
+        """
+        entries = self.auth_file.read_allow_entries()
+        for entry in entries:
+            if entry.identity == identity:
+                updated_rpc_methods = {}
+                for method in rpc_methods:
+                    updated_rpc_methods[method] = rpc_methods[method]
+                    # Check if the rpc method exists in the auth file entry
+                    if method not in entry.rpc_method_authorizations:
+                        # Create it and set it to have the provided rpc capabilities
+                        entry.rpc_method_authorizations[method] = rpc_methods[method]
+                    # Check if the rpc method does not have any rpc capabilities
+                    if not entry.rpc_method_authorizations[method]:
+                        # Set it to have the provided rpc capabilities
+                        entry.rpc_method_authorizations[method] = rpc_methods[method]
+                    # Check if the rpc method's capabilities match what have been provided
+                    if entry.rpc_method_authorizations[method] != rpc_methods[method]:
+                        # Update rpc_methods based on auth entries
+                        updated_rpc_methods[method] = entry.rpc_method_authorizations[method]
+                # Update auth file and return rpc_methods
+                self.auth_file.update_by_index(entry, entries.index(entry))
```

Comment:
Entry to the auth file should be made only if there is any change

Reply:
Fixed